Feature/speakers from firebase

fuentespatrick opened the pull request:
Same as the last PR, this isn't in the form I'd like in the long run.
But it works, which is nice

fuentespatrick committed:
Add Google Services JSON config file

fuentespatrick committed:
Add general firebase dependencies

fuentespatrick committed:
Add firebase crash reporting

fuentespatrick committed:
Update build tools

fuentespatrick committed:
Merge branch 'feature/addFirebase' into feature/scheduleFromFirebase

fuentespatrick committed:
Remove old data source and adapter references
Add TODOs to reimplement functionality once firebase data source is displaying

fuentespatrick committed:
Change speakers in the Session object to a list of strings (keys) instead of objects
Change times in the Session object from a DateTime to a String
Temporarily comment out display of speakers
Add firebase database reference and start parsing sessions from it

fuentespatrick committed:
Deprecate the data source
Comment out crashing code and add TODOs to reimplement it
Grab sessions from Firebase and populate the adapter with them
Parse the new date string format
Add documentation explaining no-arg constructor on Session model

fuentespatrick committed:
Make keys public
Remove deprecated data source from Session Details fragment
Stub fetching the session from firebase
Comment out crashing code and add related TODOs

fuentespatrick committed:
Add session ID as a member variable
Add null check for session
Fix bug in retrieving session from Firebase
Set the session member variable from Firebase

fuentespatrick committed:
Add key for getting speakers
Retrieve and display the speakers associated with a session on the Session Details screen

fuentespatrick committed:
Remove deprecated data source usage

fuentespatrick committed:
Retrieve list of speakers from Firebase

fuentespatrick committed:
Reorganize properties of the speaker model for consistency
Add image URL to the speaker model
Use the speaker image URL to pull down the image in the Speaker List

fuentespatrick committed:
Use the speaker image URL in the Speaker View

fuentespatrick committed:
Retrieve speaker from Firebase for Speaker Details

fuentespatrick committed:
Remove commented-out code

fuentespatrick committed:
Remove commented out code

fuentespatrick committed:
Remove outdated annotation

fuentespatrick committed:
Remove Dagger from the project

fuentespatrick committed:
Remove Retrofit & OkHttp from project

bherbst commented:
@fuentespatrick is this ready for review now?

fuentespatrick commented:
Yeah, I think it's fine to look at now. I'm going to abstract the firebase interaction, but I think it's reasonable to take a look at it before then. Thanks for asking, @bherbst

bherbst commented on app/google-services.json:
```diff
@@ -0,0 +1,55 @@
+{
```
Do we want this file checked into version control since it contains API keys and such?

fuentespatrick replied:
Great catch!
I can't believe that I missed that!

bherbst commented on app/src/main/java/mn/devfest/api/model/Session.java:
```diff
@@ -102,6 +113,11 @@ public String getCategory() {
+        } catch (Exception e) {
+            Timber.e(e, "Failed to parse the startTime");
+            return new DateTime();
+        }
```
Why keep the start and end times in memory as Strings instead of DateTimes when they are only used as DateTimes? If we call this method multiple times, we incur the overhead of the conversion every time.

fuentespatrick replied:
Yeah, you're absolutely correct. That's definitely an artifact of me trying to "make it work" before I "make it good".
I'll change it now

fuentespatrick replied:
Ugh. So, I originally did this because Firebase is returning strings, and I wasn't sure how to get that to map to DateTimes with getValue(). I tried changing the setter, but that doesn't seem to be working. This will probably take a few more days to fix, and I'm starting to get really concerned about my ability to get the app done before the conference.

bherbst replied:
If it is going to slow down development, #ship-it

fuentespatrick replied:
I had an idea at work today that I'm trying now.
I'm going to try just having a separate field for the DateTime and lazyloading it in it's accessor method.

bherbst commented on app/src/main/java/mn/devfest/speakers/SpeakerListFragment.java:
```diff
@@ -68,24 +69,31 @@ public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
-        }
+        mSpeakerRecyclerView.addItemDecoration(new DividerItemDecoration(getContext(), DividerItemDecoration.VERTICAL_LIST));
+        mFirebaseDatabaseReference = FirebaseDatabase.getInstance().getReference();
+        mFirebaseDatabaseReference.child(DEVFEST_2017_KEY).child(SPEAKERS_CHILD_KEY)
```
I've seen this block a few times- have you considered moving it into something like a manager/service class to eliminate the duplication?

fuentespatrick committed:
Remove Google Services from source control

fuentespatrick committed:
Prevent Google Services JSON from entering source control

fuentespatrick committed:
Change “category” to be called “track”
Add a field for start and end time to be in memory as a DateTime
Change all accessors in code to use the new accessor for DateTimes

fuentespatrick committed:
Document accessor methods

fuentespatrick committed:
Add TODOs around DateTime parsing

fuentespatrick committed:
Reference the Firebase database inside the data source class

fuentespatrick committed:
Use data source for data updates rather than directly listening to the firebase database reference
Move data source to a singleton
Move session data into the Conference field

fuentespatrick committed:
Add a method for diffing lists of sessions and updating an adapter
Override equals and hashcode for the Session model
Move Firebase access keys into the data source
Add a helper method to the Session and Speaker list fragments to handle checking for fresh data and updating the adapter
Move Speaker list fragment from directly referencing Firebase to using the Data Source

fuentespatrick committed:
Use the data source to display speakers on the Session Details screen

fuentespatrick committed:
Remove unused firebase reference from session details
Migrate Speaker Details to use the data source abstraction rather that firebase directly
